File: common/packet/example_usage.py
"""
専用パケットクラスの使用例
従来のRequest/Responseクラスと新しい専用クラスの使いやすさを比較
"""
from datetime import datetime
from .weather_packet import WeatherRequest, WeatherResponse
from .request import Request
from .response import Response
from ..clients.utils.packet_id_generator import PacketIDGenerator12Bit

# パケットIDジェネレーター
PIDG = PacketIDGenerator12Bit()


def traditional_usage_example():
    """従来のRequest/Responseクラスの使用例"""
    print("=== 従来の使用方法 ===")
    
    # 座標から天気情報を取得（従来方式）
    request = Request(
        version=1,
        packet_id=PIDG.next_id(),
        type=0,  # 座標解決リクエスト
        timestamp=int(datetime.now().timestamp()),
        weather_flag=1,
        temperature_flag=1,
        pops_flag=1,
        alert_flag=0,
        disaster_flag=0,
        day=0,
        ex_field={
            "latitude": 35.6895,
            "longitude": 139.6917
        },
        ex_flag=1
    )
    
    print("従来のRequest作成:")
    print(f"  コード行数: 15行")
    print(f"  Type: {request.type}")
    print(f"  Flags: {request.weather_flag}, {request.temperature_flag}, {request.pops_flag}")
    print(f"  Coordinates: {request.ex_field.get('latitude')}, {request.ex_field.get('longitude')}")
    
    # エリアコードから天気情報を取得（従来方式）
    request2 = Request(
        version=1,
        packet_id=PIDG.next_id(),
        type=2,  # 気象データリクエスト
        timestamp=int(datetime.now().timestamp()),
        area_code="011000",
        weather_flag=1,
        temperature_flag=1,
        pops_flag=1,
        alert_flag=1,
        disaster_flag=0,
        day=0
    )
    
    print(f"\n従来のRequest作成 (エリアコード):")
    print(f"  コード行数: 11行")
    print(f"  Type: {request2.type}")
    print(f"  Area Code: {request2.area_code}")


def modern_usage_example():
    """新しい専用クラスの使用例"""
    print("\n=== 新しい専用クラスの使用方法 ===")
    
    # 座標から天気情報を取得（新方式）
    weather_req = WeatherRequest.create_by_coordinates(
        latitude=35.6895,
        longitude=139.6917,
        packet_id=PIDG.next_id(),
        weather=True,
        temperature=True,
        precipitation_prob=True
    )
    
    print("新しいWeatherRequest作成:")
    print(f"  コード行数: 7行 (半分以下！)")
    print(f"  Type: {weather_req.type}")
    print(f"  Summary: {weather_req.get_request_summary()}")
    
    # エリアコードから天気情報を取得（新方式）
    weather_req2 = WeatherRequest.create_by_area_code(
        area_code="011000",
        packet_id=PIDG.next_id(),
        weather=True,
        temperature=True,
        precipitation_prob=True,
        alerts=True
    )
    
    print(f"\n新しいWeatherRequest作成 (エリアコード):")
    print(f"  コード行数: 7行 (従来の約半分！)")
    print(f"  Type: {weather_req2.type}")
    print(f"  Summary: {weather_req2.get_request_summary()}")


def response_processing_example():
    """レスポンス処理の比較例"""
    print("\n=== レスポンス処理の比較 ===")
    
    # サンプルレスポンスデータを作成
    sample_response = Response(
        version=1,
        packet_id=123,
        type=3,
        area_code="011000",
        timestamp=int(datetime.now().timestamp()),
        weather_flag=1,
        temperature_flag=1,
        pops_flag=1,
        alert_flag=1,
        disaster_flag=0,
        ex_flag=1,
        weather_code=100,
        temperature=125,  # 25℃ + 100
        pops=30,
        ex_field={
            'alert': ['大雨警報', '洪水注意報']
        }
    )
    
    # 従来の処理方法
    print("従来のレスポンス処理:")
    print(f"  気温: {sample_response.temperature - 100}℃ (手動変換)")
    print(f"  天気コード: {sample_response.weather_code}")
    print(f"  降水確率: {sample_response.pops}%")
    if hasattr(sample_response, 'ex_field') and sample_response.ex_field:
        alerts = sample_response.ex_field.get('alert', [])
        print(f"  警報: {alerts}")
    
    # バイト列に変換して新しいクラスで処理
    response_bytes = sample_response.to_bytes()
    weather_resp = WeatherResponse.from_bytes(response_bytes)
    
    print(f"\n新しいWeatherResponse処理:")
    print(f"  気温: {weather_resp.get_temperature_celsius()}℃ (自動変換)")
    print(f"  天気コード: {weather_resp.get_weather_code()}")
    print(f"  降水確率: {weather_resp.get_precipitation_prob_percentage()}%")
    print(f"  警報: {weather_resp.get_alerts()}")
    print(f"  成功判定: {weather_resp.is_success()}")
    print(f"  全データ: {weather_resp.get_weather_data()}")


def client_integration_example():
    """クライアント統合の例"""
    print("\n=== クライアント統合例 ===")
    
    # 新しいパケットクラスを使った簡潔なクライアント
    def create_weather_request_easily(lat, lon, options=None):
        """簡単な天気リクエスト作成"""
        options = options or {}
        return WeatherRequest.create_by_coordinates(
            latitude=lat,
            longitude=lon,
            packet_id=PIDG.next_id(),
            weather=options.get('weather', True),
            temperature=options.get('temperature', True),
            precipitation_prob=options.get('precipitation_prob', True),
            alerts=options.get('alerts', False),
            disaster=options.get('disaster', False)
        )
    
    # 使用例
    tokyo_request = create_weather_request_easily(35.6895, 139.6917)
    sapporo_request = create_weather_request_easily(
        43.0642, 141.3469, 
        {'weather': True, 'temperature': True, 'alerts': True}
    )
    
    print("簡潔なリクエスト作成:")
    print(f"  東京: {tokyo_request.get_request_summary()}")
    print(f"  札幌: {sapporo_request.get_request_summary()}")


def compatibility_test():
    """互換性テスト"""
    print("\n=== 互換性テスト ===")
    
    # 新しいクラスで作成したパケットが従来のクラスで読める
    weather_req = WeatherRequest.create_by_coordinates(
        latitude=35.6895,
        longitude=139.6917,
        packet_id=999,
        weather=True,
        temperature=True
    )
    
    # バイト列に変換
    packet_bytes = weather_req.to_bytes()
    
    # 従来のRequestクラスで読み取り
    traditional_req = Request.from_bytes(packet_bytes)
    
    print("新→従来 互換性:")
    print(f"  Type: {traditional_req.type}")
    print(f"  Coordinates: {traditional_req.ex_field.get('latitude')}, {traditional_req.ex_field.get('longitude')}")
    
    # 従来のクラスで作成したパケットが新しいクラスで読める
    old_req = Request(
        version=1,
        packet_id=888,
        type=0,
        weather_flag=1,
        temperature_flag=1,
        timestamp=int(datetime.now().timestamp()),
        ex_field={"latitude": 43.0642, "longitude": 141.3469},
        ex_flag=1
    )
    
    old_bytes = old_req.to_bytes()
    new_weather_req = WeatherRequest.from_bytes(old_bytes)
    
    print(f"\n従来→新 互換性:")
    print(f"  Summary: {new_weather_req.get_request_summary()}")


def main():
    """メイン関数"""
    print("専用パケットクラス使用例")
    print("=" * 60)
    
    traditional_usage_example()
    modern_usage_example()
    response_processing_example()
    client_integration_example()
    compatibility_test()
    
    print("\n" + "=" * 60)
    print("専用パケットクラスの利点:")
    print("✓ コード行数が大幅削減（従来の約半分）")
    print("✓ 型安全性の向上")
    print("✓ 直感的なメソッド名")
    print("✓ 自動的なデータ変換")
    print("✓ 既存コードとの完全互換性")
    print("✓ エラーの少ない開発")


if __name__ == "__main__":
    main()

File: c++common/packet/examples/ExampleUsage.cpp
#include <iostream>
#include "../../clients/QueryClient.hpp"
#include "../../clients/utils/PacketIDGenerator.hpp"
#include "../types/ReportPacket.hpp"
#include "../types/LocationPacket.hpp"
#include "../types/QueryPacket.hpp"

using namespace std;
using wip::packet::PacketIDGenerator12Bit;

// 従来方式でのパケット生成例
void traditional_usage_example(PacketIDGenerator12Bit &pidg) {
    cout << "=== 従来の使用方法 ===" << endl;

    Request request;
    request.version = 1;
    request.packet_id = pidg.next_id();
    request.type = 0; // 座標解決リクエスト
    request.timestamp = static_cast<int>(time(nullptr));
    request.weather_flag = 1;
    request.temperature_flag = 1;
    request.pop_flag = 1;
    request.alert_flag = 0;
    request.disaster_flag = 0;
    request.day = 0;
    request.ex_field["latitude"] = 35.6895;
    request.ex_field["longitude"] = 139.6917;
    request.ex_flag = 1;

    cout << "従来のRequest作成:" << endl;
    cout << "  Type: " << request.type << endl;
    cout << "  Flags: " << request.weather_flag << ", "
         << request.temperature_flag << ", " << request.pop_flag << endl;
}

// 新しい専用クラスの使用例
void modern_usage_example(PacketIDGenerator12Bit &pidg) {
    cout << "\n=== 新しい専用クラスの使用方法 ===" << endl;

    auto location_req = LocationRequest::create_coordinate_lookup(
        35.6895, 139.6917,
        pidg.next_id(),
        true,  // weather
        true,  // temperature
        true   // precipitation_prob
    );

    cout << "新しいLocationRequest作成:" << endl;
    cout << "  Type: " << location_req.type << endl;
    cout << "  Summary: " << location_req.get_request_summary() << endl;
}

// レスポンス処理の例
void response_processing_example(PacketIDGenerator12Bit &pidg) {
    cout << "\n=== レスポンス処理の比較 ===" << endl;

    Response sample_response;
    sample_response.version = 1;
    sample_response.packet_id = 123;
    sample_response.type = 3;
    sample_response.area_code = "011000";
    sample_response.timestamp = static_cast<int>(time(nullptr));
    sample_response.weather_flag = 1;
    sample_response.temperature_flag = 1;
    sample_response.pop_flag = 1;
    sample_response.alert_flag = 1;
    sample_response.disaster_flag = 0;
    sample_response.ex_flag = 1;
    sample_response.weather_code = 100;
    sample_response.temperature = 125; // 25℃ + 100
    sample_response.pop = 30;
    sample_response.ex_field["alert"] = {"大雨警報", "洪水注意報"};

    cout << "従来のレスポンス処理:" << endl;
    cout << "  気温: " << sample_response.temperature - 100 << "℃" << endl;
    cout << "  天気コード: " << sample_response.weather_code << endl;
    cout << "  降水確率: " << sample_response.pop << "%" << endl;

    auto response_bytes = sample_response.to_bytes();
    auto weather_resp = QueryResponse::from_bytes(response_bytes);

    cout << "\n新しいWeatherResponse処理:" << endl;
    cout << "  気温: " << weather_resp.get_temperature() << "℃" << endl;
    cout << "  天気コード: " << weather_resp.get_weather_code() << endl;
    cout << "  降水確率: " << weather_resp.get_precipitation_prob() << "%" << endl;
    cout << "  警報: " << weather_resp.get_alerts() << endl;
    cout << "  成功判定: " << weather_resp.is_success() << endl;
}

// クライアント統合の例
void client_integration_example(PacketIDGenerator12Bit &pidg) {
    cout << "\n=== クライアント統合例 ===" << endl;

    auto create_weather_request_easily = [&](double lat, double lon) {
        return LocationRequest::create_coordinate_lookup(
            lat,
            lon,
            pidg.next_id(),
            true,
            true,
            true
        );
    };

    auto tokyo_request = create_weather_request_easily(35.6895, 139.6917);
    auto sapporo_request = create_weather_request_easily(43.0642, 141.3469);

    cout << "簡潔なリクエスト作成:" << endl;
    cout << "  東京: " << tokyo_request.get_request_summary() << endl;
    cout << "  札幌: " << sapporo_request.get_request_summary() << endl;
}

// 互換性テスト
void compatibility_test(PacketIDGenerator12Bit &pidg) {
    cout << "\n=== 互換性テスト ===" << endl;

    auto weather_req = LocationRequest::create_coordinate_lookup(
        35.6895, 139.6917, pidg.next_id(), true, true
    );
    auto bytes = weather_req.to_bytes();
    auto traditional_req = Request::from_bytes(bytes);

    cout << "新→従来 互換性:" << endl;
    cout << "  Type: " << traditional_req.type << endl;
    auto coords = traditional_req.ex_field.get_coordinates();
    cout << "  Coordinates: " << coords.first << ", " << coords.second << endl;

    Request old_req;
    old_req.version = 1;
    old_req.packet_id = pidg.next_id();
    old_req.type = 0;
    old_req.weather_flag = 1;
    old_req.temperature_flag = 1;
    old_req.timestamp = static_cast<int>(time(nullptr));
    old_req.ex_field["latitude"] = 43.0642;
    old_req.ex_field["longitude"] = 141.3469;
    old_req.ex_flag = 1;

    auto old_bytes = old_req.to_bytes();
    auto new_weather_req = Request::from_bytes(old_bytes);

    cout << "\n従来→新 互換性:" << endl;
    cout << "  Summary: " << new_weather_req.get_request_summary() << endl;
}

int main() {
    cout << "専用パケットクラス使用例" << endl;
    cout << string(60, '=') << endl;

    PacketIDGenerator12Bit pidg;

    traditional_usage_example(pidg);
    modern_usage_example(pidg);
    response_processing_example(pidg);
    client_integration_example(pidg);
    compatibility_test(pidg);

    cout << "\n" << string(60, '=') << endl;
    cout << "専用パケットクラスの利点:" << endl;
    cout << "\xE2\x9C\x93 コード行数が大幅削減（従来の約半分）" << endl;
    cout << "\xE2\x9C\x93 型安全性の向上" << endl;
    cout << "\xE2\x9C\x93 直感的なメソッド名" << endl;
    cout << "\xE2\x9C\x93 自動的なデータ変換" << endl;
    cout << "\xE2\x9C\x93 既存コードとの完全互換性" << endl;
    cout << "\xE2\x9C\x93 エラーの少ない開発" << endl;
}

